src/gdb_packet.c

```diff
@@ -31,121 +31,163 @@
 
 #include <stdarg.h>
 
+typedef enum packet_state_e {
+	PACKET_IDLE,
+	PACKET_GDB_CAPTURE,
+	PACKET_GDB_ESCAPE,
+	PACKET_GDB_CHECKSUM_UPPER,
+	PACKET_GDB_CHECKSUM_LOWER,
+	PACKET_REMOTE_CAPTURE,
+} packet_state_e;
+
+static bool noackmode = false;
+
+void gdb_set_noackmode(bool enable)
+{
+	noackmode = enable;
+}
+
 size_t gdb_getpacket(char *const packet, const size_t size)
 {
-	unsigned char csum;
-	char recv_csum[3];
+	packet_state_e state = PACKET_IDLE; /* State of the packet capture */
 	size_t offset = 0;
+	uint8_t rx_csum = 0, csum = 0;
 
 	while (true) {
-		/* Wait for packet start */
-		do {
-			/*
-			 * Spin waiting for a start of packet character - either a gdb
-             * start ('$') or a BMP remote packet start ('!').
-			 */
-			do {
-				/* Smells like bad code */
-				packet[0] = gdb_if_getchar();
-				if (packet[0] == '\x04')
-					return 1;
-			} while (packet[0] != '$' && packet[0] != REMOTE_SOM);
+		const char c = gdb_if_getchar();
```

Comment:
We're going to be interested to see how this interacts with https://github.com/blackmagic-debug/blackmagic/tree/fix/remote-packet-handling-organisation

Reply:
ah, didn't realize there was pending work on this, could you clear up what you mean by "interacts with"? I think these are conflicting changes

Reply:
They are currently conflicting, but we're thinking if there's a way to make either that patch play nice with this, or this play nice with that.. abstracting out the remote protocol handling like that branch does is something we'd been considering a while and we think helps even with no-ack mode as the remote protocol is not and should not be influenced by no-ack mode

Reply:
ah, I follow,  I will take a look eventually, though as is no ack mode doesn't afect the remote protocol at all